src/cryptonote_core/loki_name_system.cpp

```diff
@@ -439,57 +439,75 @@ bool validate_lns_name(mapping_type type, std::string const &name, std::string *
   else
   {
     if (reason) err_stream << "LNS type=" << type << ", specifies unhandled mapping type in name validation";
     return false;
   }
 
   // NOTE: Validate name length
-  if (check_condition((name.empty() || name.size() > max_name_len), reason, "LNS type=", type, ", specifies mapping from name -> value where the name's length=", name.size(), " is 0 or exceeds the maximum length=", max_name_len, ", given name=", name))
+  if (check_condition((name.empty() || name.size() > max_name_len), reason, "LNS type=", type, ", specifies mapping from name->value where the name's length=", name.size(), " is 0 or exceeds the maximum length=", max_name_len, ", given name=", name))
     return false;
 
   // NOTE: Validate domain specific requirements
-  if (type == mapping_type::session)
+  char const *char_preceeding_suffix = nullptr;
+  if (is_lokinet)
   {
-  }
-  else if (is_lokinet_type(type))
-  {
-    // Domain has to start with a letter or digit, and can have letters, digits, or hyphens in between and must end with a .loki
+
+    // LOKINET
+    // Domain has to start with an alphanumeric, and can have (alphanumeric or hyphens) in between, the character before the suffix <char>'.loki' must be alphanumeric followed by the suffix '.loki'
     // ^[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\\.loki$
+    if (is_lokinet)
     {
+      // Must start with alphanumeric
+      if (check_condition(!char_is_alphanum(name.front()), reason, "LNS type=", type, ", specifies mapping from name->value where the name does not start with an alphanumeric character, name=", name))
+        return false;
+
       char const SHORTEST_DOMAIN[] = "a.loki";
-      if (check_condition((name.size() < static_cast<int>(loki::char_count(SHORTEST_DOMAIN))), reason, "LNS type=lokinet, specifies mapping from name -> value where the name is shorter than the shortest possible name=", SHORTEST_DOMAIN, ", given name=", name))
+      if (check_condition((name.size() < static_cast<int>(loki::char_count(SHORTEST_DOMAIN))), reason, "LNS type=", type, ", specifies mapping from name->value where the name is shorter than the shortest possible name=", SHORTEST_DOMAIN, ", given name=", name))
+        return false;
+
+      // Must end with .loki
+      char const SUFFIX[]     = ".loki";
+      char const *name_suffix = name.data() + (name.size() - loki::char_count(SUFFIX));
+      if (check_condition((memcmp(name_suffix, SUFFIX, loki::char_count(SUFFIX)) != 0), reason, "LNS type=", type, ", specifies mapping from name->value where the name does not end with the domain .loki, name=", name))
+        return false;
+
+      // Characted preceeding suffix must be alphanumeric
+      char_preceeding_suffix = name_suffix - 1;
+      if (check_condition(!char_is_alphanum(char_preceeding_suffix[0]), reason, "LNS type=", type ,", specifies mapping from name->value where the character preceeding the <char>.loki is not alphanumeric, char=", char_preceeding_suffix[0], ", name=", name))
         return false;
```

Comment:
Is this missing a check for [a-z0-9-] on the intermediate characters, or it is being handled somewhere else and I missed it?

Reply:
Oh, it's below.